contracts/adapters/midas/MidasIssuanceVaultAdapter.sol

```diff
@@ -0,0 +1,148 @@
+
+import {IMidasIssuanceVault} from "../../integrations/midas/IMidasIssuanceVault.sol";
+import {IMidasIssuanceVaultAdapter} from "../../interfaces/midas/IMidasIssuanceVaultAdapter.sol";
+
+/// @title Midas Issuance Vault adapter
+/// @notice Implements logic for interacting with the Midas Issuance Vault contract
+contract MidasIssuanceVaultAdapter is AbstractAdapter, IMidasIssuanceVaultAdapter {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    bytes32 public constant override contractType = "ADAPTER::MIDAS_ISSUANCE_VAULT";
+    uint256 public constant override version = 3_10;
+
+    /// @notice mToken
+    address public immutable override mToken;
+
+    /// @notice Referrer ID used for deposits
+    bytes32 public immutable override referrerId;
+
+    /// @dev Set of allowed input tokens for depositInstant
+    EnumerableSet.AddressSet internal _allowedTokens;
+
+    /// @notice Constructor
+    /// @param _creditManager Credit manager address
+    /// @param _midasIssuanceVault Midas Issuance Vault address
+    /// @param _referrerId Referrer ID to use for all deposits
+    constructor(address _creditManager, address _midasIssuanceVault, bytes32 _referrerId)
+        AbstractAdapter(_creditManager, _midasIssuanceVault)
+    {
+        mToken = IMidasIssuanceVault(_midasIssuanceVault).mToken();
+
+        // We check that mToken is a valid collateral
+        _getMaskOrRevert(mToken);
+
+        referrerId = _referrerId;
+    }
+
+    /// @notice Deposits specified amount of input token for mToken
+    /// @param tokenIn Input token address
+    /// @param amountToken Amount of input token to deposit
+    /// @param minReceiveAmount Minimum amount of mToken to receive
+    function depositInstant(address tokenIn, uint256 amountToken, uint256 minReceiveAmount, bytes32)
+        external
+        override
+        creditFacadeOnly
+        returns (bool)
+    {
+        if (!isTokenAllowed(tokenIn)) revert TokenNotAllowedException();
+
+        _depositInstant(tokenIn, amountToken, minReceiveAmount);
+        return false;
+    }
+
+    /// @notice Deposits entire balance of input token, except the specified amount
+    /// @param tokenIn Input token address
+    /// @param leftoverAmount Amount of input token to keep in the account
+    /// @param rateMinRAY Minimum exchange rate from input token to mToken (in RAY format)
+    function depositInstantDiff(address tokenIn, uint256 leftoverAmount, uint256 rateMinRAY)
+        external
+        override
+        creditFacadeOnly
+        returns (bool)
+    {
+        if (!isTokenAllowed(tokenIn)) revert TokenNotAllowedException();
+
+        address creditAccount = _creditAccount();
+
+        uint256 balance = IERC20(tokenIn).balanceOf(creditAccount);
+        if (balance > leftoverAmount) {
+            unchecked {
+                uint256 amount = balance - leftoverAmount;
+                uint256 minReceiveAmount = (amount * rateMinRAY) / RAY;
```

Comment:
Multiplication overflows so it should be performed outside `unchecked` block. In this case, however, there are no likely security implications (unless a user plugs in some weird `rateMinRAY` and gets sandwiched by I don't even know who).